ArduCopter/autoyaw.cpp

```diff
@@ -229,45 +227,55 @@ float Mode::AutoYaw::yaw()
         float yaw_angle_step = constrain_float(_fixed_yaw_offset_cd, - dt * _fixed_yaw_slewrate_cds, dt * _fixed_yaw_slewrate_cds);
         _fixed_yaw_offset_cd -= yaw_angle_step;
         _yaw_angle_cd += yaw_angle_step;
-        return _yaw_angle_cd;
+        break;
     }
 
     case Mode::LOOK_AHEAD:
         // Commanded Yaw to automatically look ahead.
-        return look_ahead_yaw();
+        _yaw_angle_cd = look_ahead_yaw();
+        break;
 
     case Mode::RESETTOARMEDYAW:
         // changes yaw to be same as when quad was armed
-        return copter.initial_armed_bearing;
+        _yaw_angle_cd = copter.initial_armed_bearing;
+        break;
 
     case Mode::CIRCLE:
 #if MODE_CIRCLE_ENABLED
         if (copter.circle_nav->is_active()) {
-            return copter.circle_nav->get_yaw();
+            _yaw_angle_cd = copter.circle_nav->get_yaw();
         }
 #endif
-        // return the current attitude target
-        return wrap_360_cd(copter.attitude_control->get_att_target_euler_cd().z);
```

Comment:
The removal of this line means that the _yaw_angle_cd may not be set.  The set_mode(CIRCLE) also doesn't set the yaw angle so the heading returned could be uninitialised.

A solution would be to add "_yaw_angle_cd = = copter.ahrs.yaw_sensor;" to AutoYaw::set_mode.